cli/factory.rs

```diff
@@ -0,0 +1,669 @@
+      // single executor and then we could make some initialization
+      // concurrent
+      let val = create.await?;
+      _ = self.0.set(val);
+    }
+    Ok(self.0.get().unwrap())
+  }
+}
+
+#[derive(Default)]
+struct CliFactoryServices {
+  dir: Deferred<DenoDir>,
+  caches: Deferred<Arc<Caches>>,
+  file_fetcher: Deferred<Arc<FileFetcher>>,
+  http_client: Deferred<HttpClient>,
+  emit_cache: Deferred<EmitCache>,
+  emitter: Deferred<Arc<Emitter>>,
+  graph_container: Deferred<Arc<ModuleGraphContainer>>,
+  lockfile: Deferred<Option<Arc<Mutex<Lockfile>>>>,
+  maybe_import_map: Deferred<Option<Arc<ImportMap>>>,
+  maybe_inspector_server: Deferred<Option<Arc<InspectorServer>>>,
+  root_cert_store: Deferred<RootCertStore>,
+  blob_store: Deferred<BlobStore>,
+  parsed_source_cache: Deferred<Arc<ParsedSourceCache>>,
+  resolver: Deferred<Arc<CliGraphResolver>>,
+  file_watcher: Deferred<Arc<FileWatcher>>,
+  maybe_file_watcher_reporter: Deferred<Option<FileWatcherReporter>>,
+  module_graph_builder: Deferred<Arc<ModuleGraphBuilder>>,
+  module_load_preparer: Deferred<Arc<ModuleLoadPreparer>>,
+  node_code_translator: Deferred<Arc<CliNodeCodeTranslator>>,
+  node_fs: Deferred<Arc<dyn deno_node::NodeFs>>,
+  node_resolver: Deferred<Arc<NodeResolver>>,
+  npm_api: Deferred<Arc<CliNpmRegistryApi>>,
+  npm_cache: Deferred<Arc<NpmCache>>,
+  npm_resolver: Deferred<Arc<CliNpmResolver>>,
+  npm_resolution: Deferred<Arc<NpmResolution>>,
+  package_json_deps_installer: Deferred<Arc<PackageJsonDepsInstaller>>,
+  text_only_progress_bar: Deferred<ProgressBar>,
+  type_checker: Deferred<Arc<TypeChecker>>,
+  cjs_resolutions: Deferred<Arc<CjsResolutionStore>>,
+}
+
+pub struct CliFactory {
+  maybe_sender:
+    RefCell<Option<tokio::sync::mpsc::UnboundedSender<Vec<PathBuf>>>>,
+  options: Arc<CliOptions>,
+  services: CliFactoryServices,
+}
+
+impl CliFactory {
+  pub async fn from_flags(flags: Flags) -> Result<Self, AnyError> {
+    CliFactoryBuilder::new().build_from_flags(flags).await
+  }
+
+  pub fn from_cli_options(options: Arc<CliOptions>) -> Self {
+    CliFactoryBuilder::new().build_from_cli_options(options)
+  }
+
+  pub fn cli_options(&self) -> &Arc<CliOptions> {
+    &self.options
+  }
+
+  pub fn deno_dir(&self) -> Result<&DenoDir, AnyError> {
+    self
+      .services
+      .dir
+      .get_or_try_init(|| self.options.resolve_deno_dir())
+  }
+
+  pub fn caches(&self) -> Result<&Arc<Caches>, AnyError> {
+    self.services.caches.get_or_try_init(|| {
+      let caches = Arc::new(Caches::new(self.deno_dir()?.clone()));
+      // Warm up the caches we know we'll likely need based on the CLI mode
+      match self.options.sub_command() {
+        DenoSubcommand::Run(_) => {
+          _ = caches.dep_analysis_db();
+          _ = caches.node_analysis_db();
+        }
+        DenoSubcommand::Check(_) => {
+          _ = caches.dep_analysis_db();
+          _ = caches.node_analysis_db();
+          _ = caches.type_checking_cache_db();
```

Comment:
Why is this one not initted for `run` with `--check` flag?

Reply:
This is copy and pasted from what was in ProcState before. We could improve this more for more scenarios.